src/main/java/net/minecraftforge/client/ClientCommandHandler.java

```diff
@@ -0,0 +1,312 @@
+     */
+    private static <S, T> CommandNode<T> toResult(CommandNode<S> sourceNode, Map<CommandNode<S>, CommandNode<T>> sourceToResult, S canUse, Command<T> execute,
+            Function<SuggestionProvider<S>, SuggestionProvider<T>> sourceToResultSuggestion)
+    {
+        if (sourceToResult.containsKey(sourceNode))
+            return sourceToResult.get(sourceNode);
+
+        ArgumentBuilder<T, ?> resultBuilder;
+        if (sourceNode instanceof ArgumentCommandNode<?, ?>)
+        {
+            ArgumentCommandNode<S, ?> sourceArgument = (ArgumentCommandNode<S, ?>) sourceNode;
+            RequiredArgumentBuilder<T, ?> resultArgumentBuilder = RequiredArgumentBuilder.argument(sourceArgument.getName(), sourceArgument.getType());
+            if (sourceArgument.getCustomSuggestions() != null)
+            {
+                resultArgumentBuilder.suggests(sourceToResultSuggestion.apply(sourceArgument.getCustomSuggestions()));
+            }
+            resultBuilder = resultArgumentBuilder;
+        }
+        else if (sourceNode instanceof LiteralCommandNode<?>)
+        {
+            LiteralCommandNode<S> sourceLiteral = (LiteralCommandNode<S>) sourceNode;
+            resultBuilder = LiteralArgumentBuilder.literal(sourceLiteral.getLiteral());
+        }
+        else if (sourceNode instanceof RootCommandNode<?>)
+        {
+            CommandNode<T> resultNode = new RootCommandNode<>();
+            mergeCommandNode(sourceNode, resultNode, sourceToResult, canUse, execute, sourceToResultSuggestion);
+            return resultNode;
+        }
+        else
+        {
+            throw new IllegalStateException("Node type " + sourceNode + " is not a standard node type");
+        }
+
+        if (sourceNode.getCommand() != null)
+        {
+            resultBuilder.executes(execute);
+        }
+
+        if (sourceNode.getRedirect() != null)
+        {
+            resultBuilder.redirect(toResult(sourceNode.getRedirect(), sourceToResult, canUse, execute, sourceToResultSuggestion));
+        }
+        
+        CommandNode<T> resultNode = resultBuilder.build();
+        mergeCommandNode(sourceNode, resultNode, sourceToResult, canUse, execute, sourceToResultSuggestion);
+        return resultNode;
+    }
+
+    /**
+     * Always try to execute the cached parsing of client message as a command. Requires that the execute field of the commands to be set to send to server so that they aren't
+     * treated as client command's that do nothing.
+     * 
+     * {@link net.minecraft.commands.Commands#performCommand(CommandSourceStack, String)} for reference
+     * 
+     * @param sendMessage
+     *            the chat message
+     * @return false leaves the message to be sent to the server, true means it should be caught before {@link LocalPlayer#chat(String)}
+     */
+    public static boolean sendMessage(String sendMessage)
+    {
+        ClientCommandSourceStack source = getSource();
```

Comment:
A small nitpick: This should be moved to below the `if`-check, so the source isn't constructed if not needed.